src/runtime-rs/crates/runtimes/virt_container/src/sandbox.rs

```diff
@@ -151,9 +152,51 @@ impl VirtSandbox {
             resource_configs.push(vm_rootfs);
         }
 
+        // prepare pcie port device configs
+        if let Some(port_configs) = self.prepare_pcie_port_devices().await {
+            resource_configs.extend(port_configs);
+        }
+
         Ok(resource_configs)
     }
 
+    async fn prepare_pcie_port_devices(&self) -> Option<Vec<ResourceConfig>> {
+        let mut port_configs = vec![];
+        let device_manager = self.resource_manager.get_device_manager().await;
+        let dm = device_manager.read().await;
+
+        if let Some(topo) = dm.get_pcie_topology() {
+            let (root_ports, switch_ports) = (topo.pcie_root_ports, topo.pcie_switch_ports);
+            if root_ports > 0 {
+                info!(
+                    sl!(),
+                    "prepare {} pcie root port devices for VM.", root_ports
+                );
+                let root_port_config = ResourceConfig::PortDevice(PortDeviceConfig::new(
+                    PCIePort::RootPort,
+                    root_ports,
+                ));
+
+                port_configs.push(root_port_config);
+            }
+            if switch_ports > 0 {
+                info!(
+                    sl!(),
+                    "prepare {} pcie switch port devices for VM.", switch_ports
+                );
+                let switch_port_config = ResourceConfig::PortDevice(PortDeviceConfig::new(
+                    PCIePort::SwitchPort,
+                    switch_ports,
+                ));
+
+                port_configs.push(switch_port_config);
+            }
+            Some(port_configs)
+        } else {
```

Comment:
Based on the following logic from `get_pcie_port()`:

```rust
pub fn get_pcie_port(&self) -> PCIePort {
        match (self.pcie_root_ports, self.pcie_switch_ports) {
            (0, 0) => PCIePort::NoPort,
            (r, s) if r > 0 && s > 0 => PCIePort::InvalidPort,
            (r, _) if r > 0 => PCIePort::RootPort,
            (_, s) if s > 0 => PCIePort::SwitchPort,
            _ => PCIePort::InvalidPort,
        }
    }
```
It seems a case for `r > 0 && s > 0` does not fall into this arm. I was wondering if I am missing something here.

Reply:
Here, it's used to prevent the case of both of root ports and switch ports set number which is greater than 0.
If it is removed, the (r, _) branch will contains case of (r>0, s>0), which is not allowed.
To clarify this function, test cases added to help understand.